src/index.js

```diff
@@ -0,0 +1,250 @@
+        .then(request => fetch(request))
+        .then(response => {
+          // check if response invalidates access token
+          if (config.shouldInvalidateAccessToken) {
+            return Promise.all([response, config.shouldInvalidateAccessToken(response)]);
+          }
+
+          return response;
+        })
+        .then(results => {
+          const response = results[0];
+          const shouldInvalidateAccessToken = results[1];
+
+          if (shouldInvalidateAccessToken) {
+            tokens.accessToken = null;
+            runRefreshTokenPromise();
+          }
+
+          return response;
+        })
+        .then(response => {
+          // if response is not unauthorized we don't care about it
+          if (response.status !== STATUS_UNAUTHORIZED) {
+            return response;
+          }
+
+          // if we received unauthorized and current request's token is same as access token
+          // we should refresh the token. otherwise we should just repeat request since
+          // some other request already refreshed access token
+          if (shouldFetchAccessToken(request)) {
+            runRefreshTokenPromise();
+          }
+
+          // if refresh token promise is null, it already finished before this request
+          // in that case we just want to continue and repeat this request
+          const returnPromise = refreshAccessTokenPromise || Promise.resolve();
+
+          return returnPromise
+            .then(() => {
+              // repeat request if tokens don't match
+              if (parseBearer(request.headers.get('authorization')) !== tokens.accessToken) {
+                const authorizedRequest = config.setRequestAuthorization(request, tokens.accessToken);
+
+                return fetch(authorizedRequest);
+              }
+
+              // otherwise return initial response
+              return response;
+            })
+            // fetching refresh token failed
+            .catch(error => {
+              if (error.message === ERROR_REFRESH_TOKEN_EXPIRED) {
+                outerResolve(response);
+              } else {
+                // otherwise we propagate error out
+                outerReject(error);
+              }
+            })
+        })
+        .then(outerResolve)
+        .catch(outerReject);
+
+    // if access token is not resolved yet
+    if (!tokens.accessToken) {
+      // check if we are alredy fetching it
+      if (!refreshAccessTokenPromise) {
+        // as a side-effect refreshTokenPromise is set
+        runRefreshTokenPromise();
+      }
+
+      // chain to refresh token promise (pending)
+      return refreshAccessTokenPromise
+        .then(() => runInnerPromise())
+        .catch(error => {
+          if (error.message === ERROR_INVALID_CONFIG) {
+            outerReject(error);
+          } else {
+            // if we fail to resolve refresh token, and it's not internal error
+            // just pass the request normally
```

Comment:
Here's how normal fetch behaves:
fetch(url).then(result).catch(error)
1. catch is called only on network errors (e.g. DNS unresolved)
2. then is called on all other responses, including 401

Initially I've solved this flow by throwing this exception, then capturing it in main fetch promise and resolving accordingly. In new version I did it without throwing exceptions.